### PR TITLE
storage: improve safety of merge lock

### PR DESCRIPTION
**Pull request by benesch**
Fix a known correctness bug in the merge locking scheme. See the comment
updates within for details.

As an added benefit, this fix frees store.RemoveReplica of conditionals
to deal with uninitialized replicas. Acquiring the merge lock now always
results in an initialized right-hand replica, so store.MergeRange is
always removing an initialized right-hand replica.

There are no tests for this because testing this particular edge case is
quite difficult and I'm primarily interested in avoiding the obviously
wrong call to batch.ClearRange(nil, nil) that occurs when calling
store.RemoveReplica on an unintialized replica. This nonsensical
ClearRange is, somehow, not currently an error, but it is preventing
#26872 ("engine: require iterators to specify an upper bound") from
landing.

Release note: None

---

P.S. The `storage.RemoveOptions` struct now bundles precisely one parameter. Let me know if you'd prefer I change back the signature of `store.removeReplicaInternal` to take the bool directly. Bool parameters are frowned upon, though, so I wasn't sure what was preferred.

### CODE REVIEW

**Comment by tschottdorf** [comment]
Reviewed 2 of 2 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/storage/replica.go, line 5162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-LGPd-Sc7hQioBL1GbLt:-LGPd-Sc7hQioBL1GbLu:b8wq6je) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3db5fecb740423b58bbdbcf97b007ed0a8bf7c5/pkg/storage/replica.go#L5162)):*
> ```Go
> 		// rightRepl. This adds rightRepl to the replicasByKey map so that snapshots
> 		// for overlapping keyspaces will block on its raftMu.
> 		if err := rightRepl.setDesc(&merge.RightDesc); err != nil {
> ```

I'm almost certain this is not what we want. By setting a descriptor here, you make this an initialized replica, so it's not just a placeholder -- it's a full-blown replica that thinks it's initialized (see `(*Replica).IsInitialized()`). There's maybe not a huge problem with this in the happy case because you'll immediately merge this replica away, but you definitely don't undo what you just did in the unhappy case. Plus, this Replica is likely to explode, since it's not in its own RangeDescriptor, and even more so, it might send Raft messages that possibly disrupt the peers. I really think we only want a placeholder here (or some argument that the old code is sufficient, but it's not clear to me).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27061)*

**Comment by benesch** [comment]
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/storage/replica.go, line 5162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-LGPd-Sc7hQioBL1GbLt:-LGPodgR8lFrviaycMKH:b3tt39o) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3db5fecb740423b58bbdbcf97b007ed0a8bf7c5/pkg/storage/replica.go#L5162)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I'm almost certain this is not what we want. By setting a descriptor here, you make this an initialized replica, so it's not just a placeholder -- it's a full-blown replica that thinks it's initialized (see `(*Replica).IsInitialized()`). There's maybe not a huge problem with this in the happy case because you'll immediately merge this replica away, but you definitely don't undo what you just did in the unhappy case. Plus, this Replica is likely to explode, since it's not in its own RangeDescriptor, and even more so, it might send Raft messages that possibly disrupt the peers. I really think we only want a placeholder here (or some argument that the old code is sufficient, but it's not clear to me).
</blockquote></details>

It can't send Raft messages because we hold its raftMu, right? And unless I'm mistaken the merge _cannot_ fail after this point. If it does the node crashes.

I agree that the old code was not safe.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27061)*

**Comment by tschottdorf** [comment]
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/storage/replica.go, line 5162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-LGPd-Sc7hQioBL1GbLt:-LGQUcTt1YuBdq1IQggo:b5rznqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3db5fecb740423b58bbdbcf97b007ed0a8bf7c5/pkg/storage/replica.go#L5162)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

It can't send Raft messages because we hold its raftMu, right? And unless I'm mistaken the merge _cannot_ fail after this point. If it does the node crashes.

I agree that the old code was not safe.
</blockquote></details>

It's true that the code has no way to gracefully fail after this lock is acquired, at least not today. I think you're also right about not being able to handle/send Raft messages, and it's nice that this unifies two code paths (by making the RHS replica always "exist").

But at the same time raftMu isn't what protects the descriptor, that's `r.mu`. So you're still putting a fake-initialized map into the store for the world to see. Even if this worked today and never got broken, I still think it's not clear and it will always be confusing and something to keep in mind elsewhere. We have replica placeholders for pretty much this sort of thing; I'd much prefer if we used them instead.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27061)*

**Comment by benesch** [comment]
Hmm, ok. I'll be more comfortable using the combination of a placeholder and an uninitialized replica if we can clean up the cleanup code (heh). A failed split sometimes does this:

https://github.com/cockroachdb/cockroach/blob/c094d1d07e5124024aab10645665fdcc0beef756/pkg/storage/replica.go#L5164-L5174

And sometimes this:

https://github.com/cockroachdb/cockroach/blob/c094d1d07e5124024aab10645665fdcc0beef756/pkg/storage/store.go#L2239-L2244

But then all that logic is also duplicated in https://github.com/cockroachdb/cockroach/blob/c094d1d07e5124024aab10645665fdcc0beef756/pkg/storage/store.go#L2536-L2541.

And for some reason store.removeReplicaImpl can remove a placeholder replica?? https://github.com/cockroachdb/cockroach/blob/c094d1d07e5124024aab10645665fdcc0beef756/pkg/storage/store.go#L2549 Except it blew up if you passed it a placeholder replica until I plumbed the `AllowPlaceholders` option, so why did that line exist before?

**Comment by benesch** [comment]
tl;dr is there's another PR coming your way to unify this logic first.

**Comment by benesch** [comment]
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/storage/replica.go, line 5162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-LGPd-Sc7hQioBL1GbLt:-LGVIQEy9mSQkU5ybP5O:bn37035) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3db5fecb740423b58bbdbcf97b007ed0a8bf7c5/pkg/storage/replica.go#L5162)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

It's true that the code has no way to gracefully fail after this lock is acquired, at least not today. I think you're also right about not being able to handle/send Raft messages, and it's nice that this unifies two code paths (by making the RHS replica always "exist").

But at the same time raftMu isn't what protects the descriptor, that's `r.mu`. So you're still putting a fake-initialized map into the store for the world to see. Even if this worked today and never got broken, I still think it's not clear and it will always be confusing and something to keep in mind elsewhere. We have replica placeholders for pretty much this sort of thing; I'd much prefer if we used them instead.
</blockquote></details>

Ok, reworked this. How's it look now?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27061)*

**Comment by tschottdorf** [comment]
:lgtm: though PR/commit message might've rotted (can't really see this in reviewable).

---

Reviewed 2 of 2 files at r2.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/storage/replica.go, line 5215 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-LGVt-P-2IuNzT8iaQ9F:-LGVt-P-2IuNzT8iaQ9G:bic2bib) ([raw file](https://github.com/cockroachdb/cockroach/blob/a15e5b300d238641278366d16b4c2508651174c0/pkg/storage/replica.go#L5215)):*
> ```Go
> 
> 	return func(storagebase.ReplicatedEvalResult) {
> 		rightRepl.raftMu.Unlock()
> ```

In the interest of future proofing, can you remove the placeholder if there is one (in reality, today, the merge looks like it will commit and should remove the placeholder in the process by destroying the RHS, right?).

---

*[pkg/storage/store.go, line 2373 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-LGVsg1k6UnLNqh5DLfv:-LGVsg1lDbMuPrb9475V:bntxxh9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3db5fecb740423b58bbdbcf97b007ed0a8bf7c5/pkg/storage/store.go#L2373)):*
> ```Go
> 
> // addPlaceholderLocked adds the specified placeholder. Requires that Store.mu
> // and Replica.raftMu are held.
> ```

Is there some obscure requirement here about some uninitialized replica for the given RangeID?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-:-LGVtJkMD1fugU3sQuKf:b-oz657v)*

**Comment by benesch** [comment]
Ugh, I had to rework this slightly. What I was doing before was wrong. It's much easier to handle the uninitialized case separately from the initialized case. PTAL.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/storage/replica.go, line 5215 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-LGVt-P-2IuNzT8iaQ9F:-LGWk91EBHK8NTTNMYTe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a15e5b300d238641278366d16b4c2508651174c0/pkg/storage/replica.go#L5215)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

In the interest of future proofing, can you remove the placeholder if there is one (in reality, today, the merge looks like it will commit and should remove the placeholder in the process by destroying the RHS, right?).
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 2373 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-LGVsg1k6UnLNqh5DLfv:-LGWk582DdhqFL_DAqnf:b-8jmxm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3db5fecb740423b58bbdbcf97b007ed0a8bf7c5/pkg/storage/store.go#L2373)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Is there some obscure requirement here about some uninitialized replica for the given RangeID?
</blockquote></details>

Oops, yes. https://github.com/cockroachdb/cockroach/pull/11657

I updated the comment to be far more clear. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-:-LGWkAL7B3tMONIVHJw_:b-2wu0ic)*

**Review by tbg — DISMISSED** [review]
:lgtm: just curious, what was the problem with the old code?

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-:-LGZ0V_MCg8I862MF_AI:b-5kh4qo)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica.go, line 5215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-LGZ0IyU3tNXOkba1EZG:-LGZ0IyU3tNXOkba1EZH:brk7k22) ([raw file](https://github.com/cockroachdb/cockroach/blob/779cd007b8e1da5ec4c41ca1f0a6187edcb14fd4/pkg/storage/replica.go#L5215)):*
> ```Go
> 			r.store.mu.Lock()
> 			if _, ok := r.store.mu.replicasByKey.Get(placeholder).(*ReplicaPlaceholder); ok {
> 				log.Fatalf(ctx, "merge did not remove placeholder %+v from replicasByKey", placeholder)
> ```

Just for OCD, would you Unlock here and below before Fatal'ing? This won't matter in most cases, but if the stringer for something passed into the Fatal would need the store lock, it would deadlock in a nondescript way.

**Review by benesch — COMMENTED** [review]
The way I had plumbed `opts.AllowPlaceholders` was preventing the placeholder from getting removed immediately from replicasByKey. Whoops. (It was getting removed on the next Raft tick or something, so it wasn't causing problems, but still wonky.) Anyway, it's much easier to reason about the code if uninitialized replicas never make it to removeReplicaImpl.

Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-:-LGbAvvt4fXk9OpxQwAz:b17l8xf)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica.go, line 5215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-LGZ0IyU3tNXOkba1EZG:-LGbBNCH61cwMAYNk21P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/779cd007b8e1da5ec4c41ca1f0a6187edcb14fd4/pkg/storage/replica.go#L5215)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Just for OCD, would you Unlock here and below before Fatal'ing? This won't matter in most cases, but if the stringer for something passed into the Fatal would need the store lock, it would deadlock in a nondescript way.
</blockquote></details>

Done.

**Comment by benesch** [comment]
bors r=tschottdorf

**Review by bdarnell — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27061#-:-LGbn1je4U9ZpTt-yQ9K:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

**Comment by benesch** [comment]
bors r=tschottdorf,bdarnell
